### ISSUE TITLE
Pass user's context to resolveImportMeta()

### ISSUE THREAD

**Issue by manucorporat** [rollup/rollup#2806]
- Rollup Version: 1.10.0
- Operating System (or Browser):
- Node Version:

### How Do We Reproduce?

Webpack does not support `import.meta`, in fact it makes it to crash at parse time. At stencil we "output" 4 rollup outputs:

**Browser**:
- ESM-es2017
- System (for legacy)

**To be consumed by a dependant (using webpack... or rollup):
- ESM-es5
- ESM-es2015

Currently, we use ugly find and replace hacks to replace all ocurrences of `import.meta` in the "webpack" build, since it will make the bundler to crash: https://github.com/webpack/webpack/issues/6719

The new `resolveImportMeta()` is beautiful! but there is no way to know if we are building for esm-browser, or esm-bundler, i propose to an some kind of user's-context that can be used inside the plugin hook:

```js
export function importMetaPlugin(): Plugin {
  return {
    resolveImportMeta(property: string | null, {format, userContext}) {
      if (format === 'es' && userContext && userContext.webpackBuild) {
         return doOurThing()
      }
      // fallback to default behaviour
      return null;
    },
    name: 'importMetaPlugin'
  };
}
```

The userContext could be just another option of the `generate()` method:
```js
rollupBuild.generate({
   format: 'es',
   userContext: {
     webpackBuild: true
  }
});
```

### Conclusion

We want different import.meta resolutions for the same module system, without having to run another rollup build (it's expensive).

**Comment by manucorporat** [comment]
I wish this "hook" was a property of the RollupOutputOptions, just like `dynamicImportFunction`

**Comment by lukastaegert** [comment]
So what you want is to be able to configure plugins based on the current output. I definitely see the utility here. A "context" would definitely be an easy way to implement this on Rollup side, but I am not 100% convinced if this is the best possible way of getting there.

One potential issue I see is that the context is shared between plugins without any restriction what can be put into it, which can potentially cause unwanted interactions.

Thinking about this for a while, another way of approaching it could be to make `plugins` both an input and an output option. That way instead of relying an an amorphous context (that could also not be reliably typed), now you could pass options to different plugin instances without any risk of unintended interaction.

What could this mean practically? Plugins to be used in `output.plugins` could be using the very same interface as any existing Rollup plugins. Optionally, Rollup could display a warning if a plugin used in `output.plugins` contains hooks such as `load` or `resolveId` that do not run during `generate`.

Besides your use-case, another important use I see here is creating outputs with and without minification from the same Rollup run.

Of course this would be a lot more work to refactor Rollup to support this, but it might well be worth it. What do you think?

**Comment by manucorporat** [comment]
@lukastaegert i agree with everything. The output.plugins could be a much nicer and safer solution.
I am happy to prepare a PR during this weekend if you think the idea could get into rollup!

**Comment by lukastaegert** [comment]
Definitely! This one is probably quite ambitious as I am not sure if the existing code can easily be extended, but I will support you if you need help (though probably not over the weekend)!

Some points to note:
- Plugin hooks are managed by a "plugin driver" at the moment that prepares plugin contexts (i.e. what plugins can access via `this`) for each plugin when the graph is created. I would assume that much of this code can be reused for output plugins except that additional contexts need to be prepared once per `bundle.generate` call (as internally, we will only know what output plugins we have once we get to this phase).
- There are several ways how you could play this and honestly I do not know what would work out best. If you have good ideas, just go ahead! One way is that you could still use the same pluginDriver but add a method to add output plugins at the beginning of each `generate` call that can be replaced later by the same method. The problem is that for hooks that are run during the `generate` phase such as the `resolveImportMeta` hook, it is possible that they are implemented both by input and output plugins, so they may need to be passed through both sets of plugins.
- Some of this code may seem a little hard to overlook with long functions etc. Personally, I am a fan of small and clean code, so if you find ways to extract functionality into separate helper functions etc., I would love that, but this is definitely not a must for the first approach. If there are refactoring opportunities, they can always be taken later.

**Comment by lukastaegert** [comment]
If you need feedback, you can always create a pull request marked as [WIP] where we can talk about code.

**Comment by shellscape** [comment]
@manucorporat is this something you'd still like to pursue?

**Comment by Tal500** [comment]
Any update on this?

**Comment by lukastaegert** [comment]
Output plugins are already a reality for some time. Considering the discussion, I think this issue can be closed?  https://rollupjs.org/configuration-options/#output-plugins

**Comment by Tal500** [comment]
> Output plugins are already a reality for some time. Considering the discussion, I think this issue can be closed? https://rollupjs.org/configuration-options/#output-plugins

Thanks for the info!
But why to restrict the solution only for output plugins?

**Comment by Tal500** [comment]
Clarification: For me, adding the normalized output configuration for the output hooks is enough

**Comment by lukastaegert** [comment]
That might be a misunderstanding. Nothing was added to the hooks, but as you can define plugins per output, you can pass information per output to the plugin. In any case, output options are already available in the renderStart hook.

**Comment by Tal500** [comment]
> That might be a misunderstanding. Nothing was added to the hooks, but as you can define plugins per output, you can pass information per output to the plugin. In any case, output options are already available in the renderStart hook.

Thanks for the response, but I did understand.
All I'm saying is that while output plugins are very comfortable, the user(i.e. me) would like to put the plugin logic, using the input and parallel output, in the same plugin but not to divide them into two different ones.
So my only hope was by `outputOptions` to link the actual context to it by some sort of `WeakMap`.
